### ISSUE TITLE
Improve the building of /extras

### ISSUE THREAD

**Issue by mvaligursky** [playcanvas/engine#5541]
Currently, we add engine non-core functionality to `/extras` folder, and the build system produces `playcanvas-extras.js`. This was a good initial implementation, but it is not flexible / convenient enough, as we're hoping to add more non-core functionality to extras. The possible solution here would be this, but we're open to discussion of better options as well:

- move `/extras` folder to the engine source instead of being outside, so that the new folder would be `/src/extras`
- currently the engine build is based around `/src/index.js' - both ES5 and module builds use this
- we could add additional file, perhaps called`src/index.mjs' that would be import the index.js but additionally it would import the modules from `src/extras`.
- the module build of the engine would use this `index.mjs`, and the module build of the engine would contain extras modules
- the ES5 build of the engine would use existing `index.js` and would not contain extras
- for ES5 use (in the Editor primarily), we'd build extras into individual functional modules (per subfolder in extras), so instead of playcanvas-extras.js, we'd have playcanvas-mini-stats.js and playcanvas-exporters.js as an example, allowing the user to only include what they need.

Would import maps be useful here?

**Comment by epreston** [comment]
Initial ideas:

In a world of treeshaking, having a separate "extras" only makes sense for commonjs formatted modules.  Past that, every folder, file, class, declaration is an "extra" that will be discarded / ignored / not downloaded if not referenced.

The "what doe we do with extras" is only an issue for "commonjs" and below.

**Comment by mvaligursky** [comment]
> The "what do we do with extras" is only an issue for "commonjs" and below.

Yep, fully agreed, that's the design is trying to cover.

**Comment by epreston** [comment]
If I was setting a goal for this in todays world, I would structure it for consistency / readability / ease of use in the ESM world.

Then I would use whatever clever tricks are necessary to extract commonjs build products.  Yep, import maps where its graceful in html examples that run off bundles from a CDN if needed.

things that I can see will help:

Add an exports section to the existing `package.json`  (example only)
```json
  "exports": {
    ".": {
      "import": "./build/engine.module.js",
      "require": "./build/engine.cjs"
    },
    "./examples/*": "./examples/*",
    "./src/*": "./src/*",
  },
```

A browserlist definition instead of importing out of date polyfils in `package.json`
Have rollup select and introduce the polyfils based on language and browser support targets.

```json
  "browserslist": [
    "> 1%, not dead, not ie 11, not op_mini all"
  ],
```

**Comment by mvaligursky** [comment]
> If I was setting a goal for this in todays world, I would structure it for consistency / readability / ease of use in the ESM world.

Fully agreed on this one.

The rest sounds good too, but I'm not an expert here to understand all the implications.

One other thing that'd be great would be to have the debug version of the engine published to npm in addition to current release.

**Comment by epreston** [comment]
The npm package does not contain all builds ? That would be surprising. I'll check that and get back. 

Maybe, mono-repo structure is best. It's how all the major projects are sorting this out.  If you look at it: `extras` is just the `packages` directory of a monorepo without the benefits.

The builds sort themselves out with that strategy.

**Comment by epreston** [comment]
if anyone is concerned about optimising for `bundlephobia` or "import size prior to treeshaking" or any other metric, you can configure it look good on all those metrics while including many things.

**Comment by epreston** [comment]
I might be confusing myself or just adding crazy ideas into the mix.  I got excited that you'd ask my opinion.  I'm going to write out a gross simplification of what I understand.

 - move /extras folder to /src/extras 
 - output the same build products + minor logical corrections
 - avoid any downsides or consumer changes

Reason for this, keep source with source as nature intended.  Extras might not be used by consumers of the library but the source is tightly bound to the engine due to what they are.  Simplifies the configuration of some tooling.

**Comment by epreston** [comment]
To answer your question: 

import maps are useful to: 

> end users, creating a one page html demo, who have a usage profile outside of commercial products with performance concerns, who want to use extras AFTER they have been segmented.  Quick demos, playgrounds, issue reproductions, etc.

The future world for that very specific use case would have them using "import maps" in their HTML as follows. (example only)

```html
<!-- Import maps polyfill -->
<!-- Remove this when import maps will be widely supported -->
<script async src="https://unpkg.com/es-module-shims@1.6.3/dist/es-module-shims.js">
<script type="importmap">
{
  "imports": {
    "pc": "https://code.playcanvas.com/playcanvas-stable.min.js",
    "stats": "https://code.playcanvas.com/playcanvas-mini-stats.min.js",
    "exporters": "https://code.playcanvas.com/playcanvas-exporters.min.js",
  }
}
</script>
<script type="module">
  // create a PlayCanvas application
</script>
```

In all honesty, its kinda the same as what you have now for the end user.  Just segmented slightly more.  They get an increase in performance (not downloading what they dont use) and download sizes remain consistent no matter how many extras you add.

**Comment by epreston** [comment]
Current browser support is good.  The people that need more than 84% aren't using them anyway.  They would be pulling in the new commonjs build products once segmentation is done.

See: https://caniuse.com/import-maps

**Comment by mvaligursky** [comment]
Great news! Thanks for looking at this @epreston

**Comment by epreston** [comment]
One of my projects has this structure.  So I guess that's the demo.
https://github.com/epreston/renderlayer

The extras in this case are `draco` and `gltf` from the perspective of the convenience package.  But the whole thing is modular.  I'm still gathering ideas from different projects before determining the final structure.  The idea was to move off 3 and cut builds sizes by 80% for dependent projects.

It takes advantage of the fact that the intersection between es6 / es2022 / wegl2 is so high by dropping the legacy. The straggler is driod but its not so bad, see: https://developer.android.com/about/dashboards#OpenGL

**Comment by epreston** [comment]
Questions about your original list of bullet points:

> - the ES6 build of the engine would use existing index.js and would not contain extras

you mean ES5 here, yes ?  If so, this all makes sense.

**Comment by mvaligursky** [comment]
> > * the ES6 build of the engine would use existing index.js and would not contain extras
> 
> you mean ES5 here, yes ? If so, this all makes sense.

correct. Fixed, and on the following line as well.

**Comment by epreston** [comment]
Editor needs ES5, that's an important project.  Now I understand.

You have a good plan. The build tools work best for generating ES5 if they are given ES6.  Mind if I do the first step "move extras into src" so I understand it better ?

**Comment by mvaligursky** [comment]
> Mind if I do the first step "move extras into src" so I understand it better ?

extras folder already has a build step `npm run buiid:extras` and so the folder can be moved as long as the build step keeps working

**Comment by epreston** [comment]
Did the first part. It's becoming clearer.  

Adding things to extras checklist in order:

- [ ] move files to a well named folder in extras
- [ ] modify files to import from 'playcanvas'
- [ ] add new imports / exports / re-exports to the 'index.js' in extras 
- [ ] update engine sources that reference it to import from 'playcanvas-extras'
- [ ] update unit tests
- [ ] run lint
- [ ] run unit tests
- [ ] run builds
- [ ] run build types
- [ ] run build docs

**Comment by mvaligursky** [comment]
> * update engine sources that reference it to import from 'playcanvas-extras'

engine does not reference extras, only the examples do. Same for unit tests, none for extras.
Otherwise sounds good, especially adding those to API docs. Any thoughts on this bit @willeastcott ? I think having API docs for those would be great.

**Comment by epreston** [comment]
Progress:

- [x] move /extras folder to new folder would be /src/extras

Current: --> **Nearly done, one question** <--

- [x] module builds use this additionally it would import the modules from src/extras`.
- [x] module build of the engine would contain extras modules
- [x] ES5 build of the engine would not contain extras
- [ ] do you want to bundle `fflate` or put it as a dependency of engine module builds ?

Next:

- [ ] for ES5 use (in the Editor primarily), build extras into individual functional modules (per subfolder in extras), so instead of playcanvas-extras.js, we'd have playcanvas-mini-stats.js and playcanvas-exporters.js as an example, allowing the user to only include what they need.

**Comment by epreston** [comment]
I am publishing the draft PR.

**Comment by willeastcott** [comment]
Yeah, eventually, having docs for extras would be nice, sure. Not a blocker though.

**Comment by willeastcott** [comment]
I'm not sure about `fflate`. I'd probably prefer it to be a dependency....maybe... 😄

**Comment by epreston** [comment]
It will be the first dependency for "engine".  Its an external dependency for "extras" when built by itself.  

Looks like extras gets published to npm in a broken state.  Does the documentation tell people to "npm install fflate" if they use "usdz-exporter.js" ?

**Comment by epreston** [comment]
Notes:

Changes assume checklist above is followed. The goal is everything just works without much thought.

Progress:

- [x] move /extras folder to new folder would be /src/extras
- [x] module builds import the modules from src/extras`.
- [x] module build of the engine would contain extras modules
- [x] ES5 build of the engine would not contain extras
- [x] bundle `fflate` in module builds as required by extras <-- confirm this

Current:

- [ ] create bite sized independent ES5 packages for each extra

Next: 

- [ ] ensure graceful way to consume es5 extras, potentially import maps
- [ ] ensure independent es5 packages are available on the preferred CDN
- [ ] test independent and editor environments
- [ ] update documentation

**Comment by epreston** [comment]
Some very good feedback and ideas in @kungfooman 's comment
https://github.com/playcanvas/engine/pull/5591#issuecomment-1692924887

**Comment by mvaligursky** [comment]
It would perhaps be the best before working on actual PRs to write a document with suggested design. Perhaps start with high level, and this is agreed on, add more details as needed. This could then be discussed, and PRs would be worked on when this is agreed upon by multiple interested people.

**Comment by epreston** [comment]
Where would you like these documents ?

**Comment by mvaligursky** [comment]
Perhaps a new discussion: https://github.com/playcanvas/engine/discussions
the initial post would be a document that gets updated to be current, and discussions would follow.

**Comment by mvaligursky** [comment]
hi @epreston - have you thought about this? No rush or anything, just curious.

**Comment by epreston** [comment]
Thank you for asking, yes.  Waiting on the first PR to be reviewed (might need to redo it).

I've learned a few things: `fflate` has always been a peer dependency its just not marked in the package as a peer dependency.  See `model-viewer` project.

Import maps only have a place in examples or the user scripting component of an editor.  They should only be applied when you want to create a specific developer / creator experience for consumers of this library.

I did 2 PRs, both were correct.  I closed the second one because I made mistakes with git.  It made a simple PR look huge and difficult to review.  I can redo it when the first PR is sorted.

**Comment by epreston** [comment]
You need a place to put all these extra bits that are not on the critical path.  You might only have enough time to identify things that are extras and need a place to dump them and move on.

The efforts on the new PlayCanvas component API reduce build sizes by 50%, I got a similar benefit through module segmentation.  The idea is solid.

Something graceful and well understood is what is required.  I'm already segmenting one of my projects far beyond what you need, linked above.  I want to recommend a monorepo.  I know people will love it after a few painful weeks.

**Comment by kpal81xd** [comment]
Extras now merged into engine
